lib/Dialect/TritonGPU/Transforms/AccelerateMatmul.cpp

```diff
@@ -380,6 +383,148 @@ static void decomposeMixedModeDotOp(ModuleOp mod, int computeCapability) {
+                                             rewriter.getBF16Type(), numWarps);
+    auto CTALayout = getCTALayout(oldRetType.getEncoding());
+    SmallVector<unsigned> warpsPerCTA = {numWarps, 1};
+    auto mmaEnc = NvidiaMmaEncodingAttr::get(ctx, /*versionMajor=*/versionMajor,
+                                             /*versionMinor=*/0, warpsPerCTA,
+                                             CTALayout, instrShape);
+    auto newRetType = RankedTensorType::get(
+        oldRetType.getShape(), oldRetType.getElementType(), mmaEnc);
+
+    // convert accumulator
+    auto oldAcc = dotOp.getOperand(2);
+    auto newAcc =
+        rewriter.create<ConvertLayoutOp>(oldAcc.getLoc(), newRetType, oldAcc);
+
+    auto toMMABf16 = [&newRetType, &rewriter, &ctx,
+                      &enumToType](TypedValue<RankedTensorType> v, int idx,
+                                   F8F6F4Type type) {
+      // MMAv2 Layout
+      auto vType = v.getType();
+      auto newVEncoding = DotOperandEncodingAttr::get(
+          ctx, idx, newRetType.getEncoding(), enumToType((type)));
+      auto newVType = RankedTensorType::get(
+          v.getType().getShape(), v.getType().getElementType(), newVEncoding);
+      v = rewriter.create<ConvertLayoutOp>(v.getLoc(), newVType, v);
+
+      // Bitcast
+      auto vTypeFp8 = RankedTensorType::get(
+          vType.getShape(), rewriter.getFloat8E4M3FNType(), newVEncoding);
+      v = cast<TypedValue<RankedTensorType>>(
+          rewriter.create<BitcastOp>(v.getLoc(), vTypeFp8, v).getResult());
+
+      // Convert to bf16
+      auto vTypeBf16 = RankedTensorType::get(
+          vType.getShape(), rewriter.getBF16Type(), newVEncoding);
+      return rewriter.create<FpToFpOp>(v.getLoc(), vTypeBf16, v);
+    };
+    a = toMMABf16(a, 0, aType);
+    b = toMMABf16(b, 1, bType);
+
+    // [Note: A trick to avoid warp shuffles in the lowering]
+    // FIXME: Implement this when we can set general layouts on a tensor
+
+    // For bf16, we have 4 threads per row
+    // https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#mma-16816-a-f16
+    // and each of them needs to get every scale in that row.
+    // It turns out that the layout for the output of type bf16 gives us exactly
+    // this layout when the number of mxfp vectors is equal to two (K = 64)
+    // https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#mma-16816-c
+    // This can be generalised to other K with linear layouts, but the general
+    // layout cannot cannot be represented with the predefined layouts :(
+    // With this trick, we could do the full lowering here and remove the
+    // UpcastMXFPOp altogether
+
+    auto order =
+        cast<DotOperandEncodingAttr>(a.getType().getEncoding()).getCTAOrder();
+
+    // TODO Adjust for mmav3
+    // Necessary choice to leave all the scales of the tile in that given warp
+
+    auto shapeTileA = std::array<unsigned, 2>{instrShape[0], instrShape[0]};
```

Comment:
is the second one meant to be instrShape[1]?

Reply:
No, this is a weird one. `instrShape` is always `[16, 8]` mmav2. These are the m and n values. Then, in this case, the instruction is m16n8k16, hence this weird construction.

Let me leave a comment.